server/service/src/asset/update.rs

```diff
@@ -0,0 +1,139 @@
+use super::{query::get_asset, validate::check_asset_exists};
+use crate::{service_provider::ServiceContext, NullableUpdate, SingleRecordError};
+use chrono::{NaiveDate, Utc};
+use repository::{
+    assets::{
+        asset::{AssetFilter, AssetRepository},
+        asset_row::{AssetRow, AssetRowRepository},
+    },
+    EqualFilter, RepositoryError, StorageConnection, StringFilter,
+};
+
+#[derive(PartialEq, Debug)]
+pub enum UpdateAssetError {
+    AssetDoesNotExist,
+    AssetDoesNotBelongToCurrentStore,
+    SerialNumberAlreadyExists,
+    UpdatedRecordNotFound,
+    DatabaseError(RepositoryError),
+}
+
+#[derive(Debug, Default)]
+pub struct UpdateAsset {
+    pub id: String,
+    pub store_id: Option<String>,
+    pub name: Option<String>,
+    pub code: Option<String>,
+    pub serial_number: Option<NullableUpdate<String>>,
+    pub catalogue_item_id: Option<NullableUpdate<String>>,
+    pub installation_date: Option<NullableUpdate<NaiveDate>>,
+    pub replacement_date: Option<NullableUpdate<NaiveDate>>,
+}
+
+pub fn update_asset(
+    ctx: &ServiceContext,
+    input: UpdateAsset,
+) -> Result<AssetRow, UpdateAssetError> {
+    let asset = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let asset_row = validate(connection, &ctx.store_id, &input)?;
+            let updated_asset_row = generate(&ctx.store_id, input, asset_row);
+            AssetRowRepository::new(&connection).upsert_one(&updated_asset_row)?;
+
+            get_asset(ctx, updated_asset_row.id).map_err(UpdateAssetError::from)
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(asset)
+}
+
+pub fn validate(
+    connection: &StorageConnection,
+    ctx_store_id: &str,
+    input: &UpdateAsset,
+) -> Result<AssetRow, UpdateAssetError> {
+    let asset_row = match check_asset_exists(&input.id, connection)? {
+        Some(asset_row) => asset_row,
+        None => return Err(UpdateAssetError::AssetDoesNotExist),
+    };
+
+    // Check the serial number is unique (if present)
+    if let Some(serial_number) = &input.serial_number {
+        if let Some(serial_number) = &serial_number.value {
+            if AssetRepository::new(connection)
+                .query_one(
+                    AssetFilter::new()
+                        .id(EqualFilter::not_equal_to(&asset_row.id))
+                        .serial_number(StringFilter::equal_to(serial_number)),
+                )?
+                .is_some()
+            {
+                return Err(UpdateAssetError::SerialNumberAlreadyExists);
+            }
+        }
+    }
+    if let Some(store_id) = &asset_row.store_id {
+        // TODO: confirm, maybe people can just create them on central for any store
```

Comment:
Question, I've seen this pattern in things I'm copying, but do we want to remove this restriction? What happens for non-store items, do we need to check we are on central?

Reply:
given how we are using store as a proxy for 'where is the asset located?' I think remove the restriction. that was assets can be created centrally, and not assigned to a store. Of course we could go both ways - require store if you are on a remote site, and allow no store if on central. easier to just remove the restriction